internal/charmstore/addentity.go

```diff
@@ -562,16 +562,30 @@ func (s *Store) addCharm(c charm.Charm, p addParams) (err error) {
 // setEntityChannels associates the entity with the given channels, ignoring
 // unknown channels.
 func setEntityChannels(entity *mongodoc.Entity, chans []params.Channel) {
+	entity.Published = make(map[params.Channel]bool, len(chans))
 	for _, c := range chans {
-		switch c {
-		case params.EdgeChannel:
-			entity.Edge = true
-		case params.StableChannel:
-			entity.Stable = true
+		if ValidChannels[c] {
+			entity.Published[c] = true
 		}
 	}
 }
 
+// OrderedChannels holds the list of valid channels in order of publishing
+// status, most stable first.
+var OrderedChannels = []params.Channel{
```

Comment:
If we had UnpublishedChannel here, then we could avoid defining a separate ValidChannels variable in v5, I think.